articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,225 @@
+
+5. Adapt the ROS 2 API or underlying libraries to implement these solutions and to improve the experience of writing real-time code.
+
+6. Refactor the system for the new API if necessary and repeat.
+
+## Software architecture challenges and solutions
+
+Wherever ROS 2 libraries expose sources of nondeterminism, there should be a path that exposes an alternative.
+This section evaluates the current state of ROS 2 and rclcpp, predicts challenges to fulfilling real-time requirements, and presents possible solutions.
+
+### Memory Allocation
+
+Poor memory management can create performance issues due to page faults and nondeterministic algorithms for dynamically allocating memory.
+Although C++ has an advantage over many programming languages because memory management is exposed, the default memory allocator for operator `new` and dynamically sized data structures is not well-suited to real-time applications (see [Composing High-Performance Memory Allocators](https://people.cs.umass.edu/~emery/pubs/berger-pldi2001.pdf)).
+C++11 implements an allocator interface for injecting a custom allocator into standard library structures such as `std::vector` (see [cppreference.com](http://en.cppreference.com/w/cpp/concept/Allocator)).
+To allow for optimal performance in `rclcpp`, the client library should expose custom allocators for publishers, subscribers, etc. that are propagated to standard library structures used by these entities.
+
+As of the writing of this document (1/6/16), selecting custom allocators in `rclcpp` is almost fully implemented.
+An example of the ROS 2 allocator API using the Two Level Segregate Fit allocator is available [on Github](https://github.com/ros2/realtime_support/blob/master/tlsf_cpp/example/allocator_example.cpp).
+
+The following work remains to be done:
+
+- Improve the interface in rclcpp such that the user need only provide an "allocate" function, a "deallocate" function, and an optional pointer to the allocator state (this interface is currently outlined in the rcl API).
+- Custom allocators for services/clients.
+- Custom allocators for parameters.
+- Fully implement and test the allocator interface in rcl, rclc, and rmw.
+
+### Scheduling and synchronization
+
+Parallelism is valuable for real-time systems because it can reduce latency.
+
+Even in single-core architectures, concurrency is important to us because asynchronous patterns are common in ROS and ROS 2.
+
+The typical approaches to synchronizing concurrent access to data tend to decrease the determinism of a system and increase the complexity.
+For example, mutexes introduce the possibility of deadlocks when used improperly.
+As mentioned above, many real-time operating systems offer alternative synchronization primitives for real-time safety.
+
+Guidelines for synchronization in ROS 2:
+
+- Prefer atomics to mutexes when possible.
```

Comment:
Might be worth mentioning that atomics should be checkable to see if they are lock-free. Therefore our code will be lock-free given these other conditions hold, e.g. `atomic_is_lock_free<T>` is `true` for `T`.